### ISSUE TITLE
IndexPath assertion fails when dealing with more than one section in ASTableView

### ISSUE THREAD

**Issue by pNre** [facebook/AsyncDisplayKit#321]
Since the PR #311, using an `ASTableView` with more than one section causes the assertion in `ASFindElementsInMultidimensionalArrayAtIndexPaths` to fail with an:

> 'NSInternalInconsistencyException', reason: 'Indexpath is out of range !'

Looks like `curIdx` only accounts for the rows in the first section and when `indexPaths` contains paths for more than a section `curIdx == indexPaths.count` is always false.

**Comment by tanliboy** [comment]
@pNre , thanks for report !
The `curIdx` is a counter for the elements that has been affected in the multi-dimensional array. I tried to insert indexPath from multiple sections but I could not repro the bug. Have you inserted the section before inserting the indexPath ? The assertion is to catch invalid function call in that case.

**Comment by pNre** [comment]
I tested this in the *Kittens* example project.

I just added a section, as if I was dealing with a `UITableView`:

```objc
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
  if (section == 1) { return 1; }
  // blurb node + kLitterSize kitties
  return 1 + _kittenDataSource.count;
}


- (ASCellNode *)tableView:(ASTableView *)tableView nodeForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.section == 1) {
        return [[KittenNode alloc] initWithKittenOfSize:CGSizeMake(w, h)];
    }

// and so on ....

```

Scrolling to the second section causes the assertion to fail.

**Comment by lxcid** [comment]
+1, I just faced similar issue. Will try to find a repro as well.

**Comment by lxcid** [comment]
Confirm that @pNre sample do have the same repro.

**Comment by lxcid** [comment]
In `ASMultidimensionalArrayUtils.mm`, the old logic for the following function, while iterating through the `indexPaths` array, break early as long as the the section of the index path in question does not match the current section (represented by `curIndexPath`). This can cause problem when `indexPaths` array contains more than 1 section worth of index paths. Also, `indexPaths` array is not in sorted order. (See https://www.dropbox.com/s/7ufdcnig1v3e3a4/Screenshot%202015-02-24%2006.57.41.png?dl=0)

The follow code is a prototype that fix the issue. The only code I omitted is the defensive check `curIdx < indexPaths.count`. I can't think of a way that the logic will break the check, but I don't have full context to be able to make such assumption with confidence.

My logic isn't very efficient. as for each section, it loops through the whole `indexPaths` array. It work well in the situation as the logic does not need the `indexPaths` to be sorted. Practically, as the range returned is only a small subset of the total, I might be able to get away with this. :P

A better logic could emerge if we sorted the `indexPaths` array through `compare:` but I'm opting for simpler implementation. You guys probably have a better approach.

I hope this help! :)

```Objective-C
static void ASRecursivelyUpdateMultidimensionalArrayAtIndexPaths(NSMutableArray *mutableArray,
                                                                 const NSArray *indexPaths,
                                                                 NSUInteger &curIdx,
                                                                 NSIndexPath *curIndexPath,
                                                                 const NSUInteger dimension,
                                                                 void (^updateBlock)(NSMutableArray *arr, NSIndexSet *indexSet, NSUInteger idx)) {
  if (curIdx == indexPaths.count) {
    return;
  }

  if (curIndexPath.length < dimension - 1) {
    for (int i = 0; i < mutableArray.count; i++) {
      ASRecursivelyUpdateMultidimensionalArrayAtIndexPaths(mutableArray[i], indexPaths, curIdx, [curIndexPath indexPathByAddingIndex:i], dimension, updateBlock);
    }
  } else {
    NSMutableIndexSet *indexSet = [[NSMutableIndexSet alloc] init];

    // Old logic
//    while (curIdx < indexPaths.count &&
//           [curIndexPath isEqual:[indexPaths[curIdx] indexPathByRemovingLastIndex]]) {
//      [indexSet addIndex:[indexPaths[curIdx] indexAtPosition:curIndexPath.length]];
//      curIdx++;
//    }

    // New logic
    for (NSIndexPath *indexPath in indexPaths) {
      NSIndexPath *sectionIndexPath = [indexPath indexPathByRemovingLastIndex];
      if ([curIndexPath isEqual:sectionIndexPath]) {
        NSUInteger index = [indexPath indexAtPosition:curIndexPath.length];
        [indexSet addIndex:index];
        curIdx++;
      }
    }

    updateBlock(mutableArray, indexSet, curIdx);
  }
}
```

**Comment by tanliboy** [comment]
Oh, thanks a lot !
The function assume passed indexPaths should be sorted, not sure why it is not sorted. I will dig into a moment later.

**Comment by lxcid** [comment]
Ahhh, I just realise that. Cool. :D

No problem man! Thanks!

**Comment by tanliboy** [comment]
Thanks all for catching it ! It is a serious bug, and sorry about it.
Send a diff to fix it.

**Comment by secretiverhyme** [comment]
Fixed in #325.
